File: backend/src/main/java/com/agentworkbench/auth/service/LdapAuthService.java
package com.agentworkbench.auth.service;

import com.agentworkbench.auth.controller.AuthController.*;
import com.agentworkbench.common.exception.BusinessException;
import com.agentworkbench.permission.entity.UserRole;
import com.agentworkbench.user.entity.User;
import com.agentworkbench.user.mapper.UserMapper;
import com.agentworkbench.permission.mapper.UserRoleMapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.naming.Context;
import javax.naming.NamingEnumeration;
import javax.naming.NamingException;
import javax.naming.directory.*;
import javax.naming.ldap.InitialLdapContext;
import javax.naming.ldap.LdapContext;
import java.time.LocalDateTime;
import java.util.Hashtable;

@Slf4j
@Service
@RequiredArgsConstructor
public class LdapAuthService {

    private final UserMapper userMapper;
    private final UserRoleMapper userRoleMapper;
    private final JwtService jwtService;

    @Value("${ldap.url:}")
    private String ldapUrl;

    @Value("${ldap.base-dn:}")
    private String baseDn;

    @Value("${ldap.user-dn:}")
    private String userDn;

    @Value("${ldap.password:}")
    private String ldapPassword;

    @Value("${ldap.user-search-base:ou=users}")
    private String userSearchBase;

    public LoginVO login(String username, String password) {
        if (ldapUrl == null || ldapUrl.isEmpty()) {
            throw new BusinessException(5003, "LDAP 未配置");
        }

        try {
            // Step 1: Bind to LDAP with admin credentials
            LdapContext adminCtx = createLdapContext();

            // Step 2: Search for user
            String userDnPath = findUserDn(adminCtx, username);
            adminCtx.close();

            if (userDnPath == null) {
                throw new BusinessException(5004, "LDAP 用户不存在");
            }

            // Step 3: Authenticate user
            LdapContext userCtx = createLdapContext(userDnPath, password);
            Attributes attrs = userCtx.getAttributes(userDnPath);
            userCtx.close();

            String displayName = getAttr(attrs, "cn", username);
            String email = getAttr(attrs, "mail", null);

            // Step 4: Find or create local user
            User user = userMapper.selectOne(
                    new QueryWrapper<User>().eq("username", username).eq("auth_type", "LDAP"));

            if (user == null) {
                user = new User();
                user.setUsername(username);
                user.setDisplayName(displayName);
                user.setEmail(email);
                user.setAuthType("LDAP");
                user.setStatus(1);
                userMapper.insert(user);

                // Assign default USER role
                UserRole ur = new UserRole();
                ur.setUserId(user.getId());
                ur.setRoleId(2L);
                userRoleMapper.insert(ur);
            } else {
                user.setDisplayName(displayName);
                user.setEmail(email);
                user.setLastLoginAt(LocalDateTime.now());
                userMapper.updateById(user);
            }

            // Step 5: Generate JWT tokens
            String accessToken = jwtService.generateToken(user.getId(), user.getUsername());
            String refreshToken = jwtService.generateRefreshToken(user.getId(), user.getUsername());

            LoginVO vo = new LoginVO();
            vo.setAccessToken(accessToken);
            vo.setRefreshToken(refreshToken);
            vo.setExpiresIn(86400L);

            UserInfoVO userInfo = new UserInfoVO();
            userInfo.setId(user.getId());
            userInfo.setUsername(user.getUsername());
            userInfo.setDisplayName(user.getDisplayName());
            userInfo.setEmail(user.getEmail());
            userInfo.setAvatarUrl(user.getAvatarUrl());
            vo.setUser(userInfo);

            return vo;

        } catch (NamingException e) {
            log.error("LDAP authentication failed for user: {}", username, e);
            throw new BusinessException(5004, "LDAP 认证失败: " + e.getMessage());
        }
    }

    private LdapContext createLdapContext() throws NamingException {
        return createLdapContext(userDn, ldapPassword);
    }

    private LdapContext createLdapContext(String bindDn, String password) throws NamingException {
        Hashtable<String, String> env = new Hashtable<>();
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
        env.put(Context.PROVIDER_URL, ldapUrl);
        env.put(Context.SECURITY_AUTHENTICATION, "simple");
        env.put(Context.SECURITY_PRINCIPAL, bindDn);
        env.put(Context.SECURITY_CREDENTIALS, password);
        env.put(Context.REFERRAL, "follow");
        return new InitialLdapContext(env, null);
    }

    private String findUserDn(LdapContext ctx, String username) throws NamingException {
        String searchBase = userSearchBase + "," + baseDn;
        SearchControls controls = new SearchControls();
        controls.setSearchScope(SearchControls.SUBTREE_SCOPE);
        controls.setReturningAttributes(new String[]{"dn", "cn", "mail"});

        NamingEnumeration<SearchResult> results = ctx.search(
                searchBase, "(uid={0})", new Object[]{username}, controls);

        if (results.hasMore()) {
            SearchResult result = results.next();
            return result.getNameInNamespace();
        }
        return null;
    }

    private String getAttr(Attributes attrs, String name, String defaultVal) {
        try {
            Attribute attr = attrs.get(name);
            if (attr != null && attr.size() > 0) {
                return attr.get(0).toString();
            }
        } catch (NamingException ignored) {
        }
        return defaultVal;
    }
}

File: backend-ts/src/auth/ldap-auth.service.ts
import { Client } from 'ldapts';
import { BusinessException } from '../common/business-exception.js';
import { JwtService } from '../crypto/jwt.service.js';
import type { AppConfig } from '../config/app-config.js';
import { hasText } from '../common/case.js';
import type { LoginVO, User, UserRepository, UserRoleRepository } from '../user/types.js';
import { UserService } from '../user/user.service.js';
import { formatNow } from './auth.service.js';

export class LdapAuthService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly userRoleRepo: UserRoleRepository,
    private readonly jwtService: JwtService,
    private readonly cfg: AppConfig['ldap'],
  ) {}

  isConfigured(): boolean {
    return this.cfg.enabled && hasText(this.cfg.url);
  }

  async authenticate(username: string, password: string): Promise<LoginVO> {
    if (!this.isConfigured()) {
      throw new BusinessException(5003, 'LDAP 未配置');
    }
    const admin = new Client({ url: this.cfg.url });
    try {
      await admin.bind(this.cfg.userDn, this.cfg.password);
      const searchBase = `${this.cfg.userSearchBase},${this.cfg.baseDn}`;
      const { searchEntries } = await admin.search(searchBase, {
        scope: 'sub',
        filter: `(sAMAccountName=${escapeFilter(username)})`,
        attributes: ['dn', 'cn', 'mail'],
      });
      await admin.unbind();
      if (searchEntries.length === 0) {
        throw new BusinessException(5004, 'LDAP 用户不存在');
      }
      const entry = searchEntries[0];
      const userDnPath = String(entry.dn);
      const userClient = new Client({ url: this.cfg.url });
      try {
        await userClient.bind(userDnPath, password);
        const displayName = attr(entry, 'cn') ?? username;
        const email = attr(entry, 'mail');
        let user = await this.userRepo.findByUsername(username);
        if (!user) {
          user = {
            username,
            displayName,
            email,
            status: 1,
          };
          await this.userRepo.insert(user);
          await this.userRoleRepo.insert({ userId: user.id!, roleId: 2 });
        } else {
          user.displayName = displayName;
          user.email = email;
          user.lastLoginAt = formatNow();
          await this.userRepo.updateById(user);
        }
        return {
          accessToken: this.jwtService.generateToken(user.id!, user.username),
          refreshToken: this.jwtService.generateRefreshToken(user.id!, user.username),
          expiresIn: 86400,
          user: {
            id: user.id,
            username: user.username,
            displayName: user.displayName,
            email: user.email,
            avatarUrl: user.avatarUrl,
            authSource: UserService.resolveAuthSource(user),
          },
        };
      } finally {
        try {
          await userClient.unbind();
        } catch {
          /* ignore */
        }
      }
    } catch (e) {
      if (e instanceof BusinessException) {
        throw e;
      }
      console.error(`LDAP authentication failed for user: ${username}`, e);
      throw new BusinessException(5004, `LDAP 认证失败: ${e instanceof Error ? e.message : String(e)}`);
    }
  }
}

function attr(entry: Record<string, unknown>, name: string): string | undefined {
  const v = entry[name];
  if (Array.isArray(v)) {
    return v[0] != null ? String(v[0]) : undefined;
  }
  return v != null ? String(v) : undefined;
}

function escapeFilter(value: string): string {
  return value.replace(/[\\*()\0]/g, (c) => {
    const map: Record<string, string> = { '\\': '\\5c', '*': '\\2a', '(': '\\28', ')': '\\29', '\0': '\\00' };
    return map[c] ?? c;
  });
}
